File: packages/agent-core/src/ports.ts
export type SpiritAgentMode = 'agent' | 'plan' | 'ask' | 'debug';

export function normalizeSpiritAgentMode(input?: {
  agentMode?: unknown;
  planMode?: boolean;
}): SpiritAgentMode {
  if (input?.agentMode === 'agent' || input?.agentMode === 'plan' || input?.agentMode === 'ask' || input?.agentMode === 'debug') {
    return input.agentMode;
  }
  return input?.planMode === true ? 'plan' : 'agent';
}

export function readSpiritAgentModeFromTransportConfig(
  config: { spiritAgentMode?: SpiritAgentMode; planMode?: boolean } | undefined,
): SpiritAgentMode {
  if (config?.spiritAgentMode === 'agent' || config?.spiritAgentMode === 'plan' || config?.spiritAgentMode === 'ask' || config?.spiritAgentMode === 'debug') {
    return config.spiritAgentMode;
  }
  return config?.planMode === true ? 'plan' : 'agent';
}

export type JsonPrimitive = string | number | boolean | null;
export type JsonValue = JsonPrimitive | JsonObject | JsonValue[];

export interface JsonObject {
  [key: string]: JsonValue;
}

export type ChatRole = 'system' | 'user' | 'assistant' | 'tool';

export interface LlmToolCall {
  id: string;
  name: string;
  argumentsJson: string;
}

export interface LlmTextContentPart extends JsonObject {
  type: 'text';
  text: string;
}

export interface LlmImageContentPart extends JsonObject {
  type: 'image';
  path: string;
}

export interface LlmVideoContentPart extends JsonObject {
  type: 'video';
  path: string;
}

export type LlmContentPart = LlmTextContentPart | LlmImageContentPart | LlmVideoContentPart;
export type LlmMessageContent = LlmContentPart[];

export interface LlmMessage {
  role: ChatRole;
  content: LlmMessageContent;
  toolCallId?: string;
  toolCalls?: LlmToolCall[];
  providerState?: JsonObject;
}

export interface LegacyLlmMessageArchiveEntry {
  role: ChatRole;
  content: string;
  imagePaths?: string[];
  toolCallId?: string;
  toolCalls?: LlmToolCall[];
  providerState?: JsonObject;
}

export interface StoredLlmMessageArchiveEntry {
  role: ChatRole;
  content: LlmMessageContent;
  toolCallId?: string;
  toolCalls?: LlmToolCall[];
  providerState?: JsonObject;
}

function cloneLlmToolCalls(toolCalls: readonly LlmToolCall[]): LlmToolCall[] {
  return toolCalls.map((toolCall) => ({
    id: toolCall.id,
    name: toolCall.name,
    argumentsJson: toolCall.argumentsJson,
  }));
}

export function cloneLlmProviderState(providerState: JsonObject): JsonObject {
  return Object.fromEntries(
    Object.entries(providerState).map(([key, value]) => [key, cloneJsonValue(value)]),
  );
}

function cloneJsonValue(value: JsonValue): JsonValue {
  if (Array.isArray(value)) {
    return value.map((item) => cloneJsonValue(item));
  }
  if (typeof value === 'object' && value !== null) {
    return cloneLlmProviderState(value);
  }
  return value;
}

export interface ToolExecutionOutput {
  content: LlmMessageContent;
  summaryText: string;
  hostUi?: ToolExecutionHostUi;
}

export interface ToolExecutionHostUi {
  lspWriteDiagnostics?: import('./lsp/types.js').LspWriteDiagnosticsUi;
}

export const DEFAULT_IMAGE_GENERATION_SIZE = '1024x1024';
export const DEFAULT_VIDEO_GENERATION_DURATION = 5;

export interface ImageGenerationRequest {
  prompt: string;
  size: string;
}

export interface VideoGenerationRequest {
  prompt: string;
  duration?: number;
  aspectRatio?: string;
  resolution?: string;
}

export interface GeneratedImageSaveRequest {
  data: Uint8Array;
  mediaType: string;
  prompt: string;
  model: string;
}

export interface GeneratedImageFile {
  path: string;
  mimeType: string;
  markdownRef: string;
}

export interface GeneratedVideoSaveRequest {
  data: Uint8Array;
  mediaType: string;
  prompt: string;
  model: string;
}

export interface GeneratedVideoFile {
  path: string;
  mimeType: string;
  markdownRef: string;
}

export function createToolExecutionTextOutput(text: string): ToolExecutionOutput {
  return {
    content: createLlmMessageContentFromText(text),
    summaryText: text,
  };
}

export function createLlmTextContentPart(text: string): LlmTextContentPart {
  return { type: 'text', text };
}

export function createLlmImageContentPart(path: string): LlmImageContentPart {
  return { type: 'image', path };
}

export function createLlmVideoContentPart(path: string): LlmVideoContentPart {
  return { type: 'video', path };
}

export function createLlmMessageContentFromText(text: string): LlmMessageContent {
  return text.length > 0 ? [createLlmTextContentPart(text)] : [];
}

export function createLlmMessageContentFromTextAndImages(
  text: string,
  imagePaths: readonly string[] = [],
  videoPaths: readonly string[] = [],
): LlmMessageContent {
  const content: LlmMessageContent = [];
  if (text.length > 0) {
    content.push(createLlmTextContentPart(text));
  }
  for (const imagePath of imagePaths) {
    content.push(createLlmImageContentPart(imagePath));
  }
  for (const videoPath of videoPaths) {
    content.push(createLlmVideoContentPart(videoPath));
  }
  return content;
}

export function cloneLlmMessageContent(content: readonly LlmContentPart[]): LlmMessageContent {
  return content.map((part) => {
    if (part.type === 'text') {
      return { type: 'text', text: part.text };
    }

    if (part.type === 'video') {
      return { type: 'video', path: part.path };
    }

    return { type: 'image', path: part.path };
  });
}

export function llmMessageTextContent(content: readonly LlmContentPart[]): string {
  return content
    .filter((part): part is LlmTextContentPart => part.type === 'text')
    .map((part) => part.text)
    .join('');
}

export function llmMessageImagePaths(content: readonly LlmContentPart[]): string[] {
  return content
    .filter((part): part is LlmImageContentPart => part.type === 'image')
    .map((part) => part.path);
}

export function llmMessageHasImages(content: readonly LlmContentPart[]): boolean {
  return content.some((part) => part.type === 'image');
}

export function llmMessageVideoPaths(content: readonly LlmContentPart[]): string[] {
  return content
    .filter((part): part is LlmVideoContentPart => part.type === 'video')
    .map((part) => part.path);
}

export function llmMessageHasVideos(content: readonly LlmContentPart[]): boolean {
  return content.some((part) => part.type === 'video');
}

export function llmMessageHasMedia(content: readonly LlmContentPart[]): boolean {
  return llmMessageHasImages(content) || llmMessageHasVideos(content);
}

export function llmMessageHasText(content: readonly LlmContentPart[]): boolean {
  return content.some((part) => part.type === 'text' && part.text.trim().length > 0);
}

export function llmMessageContentWithoutImages(
  content: readonly LlmContentPart[],
): LlmMessageContent {
  return content
    .filter((part): part is LlmTextContentPart => part.type === 'text')
    .map((part) => ({ type: 'text', text: part.text }));
}

export function normalizeStoredLlmMessage(
  message: StoredLlmMessageArchiveEntry | LegacyLlmMessageArchiveEntry,
): LlmMessage {
  const toolCallId =
    'toolCallId' in message && typeof message.toolCallId === 'string'
      ? message.toolCallId
      : 'tool_call_id' in message && typeof message.tool_call_id === 'string'
        ? message.tool_call_id
        : undefined;
  const toolCalls =
    'toolCalls' in message && Array.isArray(message.toolCalls)
      ? cloneLlmToolCalls(message.toolCalls)
      : 'tool_calls' in message && Array.isArray(message.tool_calls)
        ? cloneLlmToolCalls(message.tool_calls)
        : undefined;
  const providerState =
    'providerState' in message && typeof message.providerState === 'object' && message.providerState !== null
      ? cloneLlmProviderState(message.providerState)
      : 'provider_state' in message
          && typeof message.provider_state === 'object'
          && message.provider_state !== null
        ? cloneLlmProviderState(message.provider_state as JsonObject)
        : undefined;

  if (Array.isArray(message.content)) {
    return {
      role: message.role,
      content: cloneLlmMessageContent(message.content),
      ...(toolCallId !== undefined ? { toolCallId } : {}),
      ...(toolCalls !== undefined ? { toolCalls } : {}),
      ...(providerState !== undefined ? { providerState } : {}),
    };
  }

  return {
    role: message.role,
    content: createLlmMessageContentFromTextAndImages(
      message.content,
      'imagePaths' in message ? message.imagePaths ?? [] : [],
    ),
    ...(toolCallId !== undefined ? { toolCallId } : {}),
    ...(toolCalls !== undefined ? { toolCalls } : {}),
    ...(providerState !== undefined ? { providerState } : {}),
  };
}

export interface AssistantAuxArchiveEntry {
  messageIndex: number;
  thinking?: string;
  compaction?: string;
  finishTaskNotice?: string;
}

export type SubagentSessionStatus = 'running' | 'completed' | 'failed' | 'blocked';

export interface SubagentSessionSummary {
  sessionId: string;
  parentToolCallId: string;
  title: string;
  status: SubagentSessionStatus;
  startedAtUnixMs: number;
  updatedAtUnixMs: number;
  completedAtUnixMs?: number;
  latestMessage?: string;
  finalOutput?: string;
  error?: string;
}

export interface SubagentSessionArchiveEntry {
  summary: SubagentSessionSummary;
  llmHistory: Array<StoredLlmMessageArchiveEntry | LegacyLlmMessageArchiveEntry>;
}

export interface DreamScope {
  workspaceRoot: string;
  gitBranch: string;
}

export type DreamRecordStatus = 'active' | 'superseded' | 'deleted';

export interface DreamSourceSessionRef {
  path: string;
  displayName?: string;
  savedAtUnixMs?: number;
}

export interface DreamRecord {
  id: string;
  scope: DreamScope;
  title: string;
  summary: string;
  details?: string;
  tags?: string[];
  status: DreamRecordStatus;
  createdAtUnixMs: number;
  updatedAtUnixMs: number;
  expiresAtUnixMs: number;
  sourceSessions: DreamSourceSessionRef[];
}

export interface DreamQuery {
  scope: DreamScope;
  nowUnixMs?: number;
  includeExpired?: boolean;
  includeDeleted?: boolean;
}

export type DreamCollectorRunStatus =
  | 'queued'
  | 'running'
  | 'completed'
  | 'failed'
  | 'skipped'
  | 'backoff';

export type DreamCollectorDecision = 'created' | 'updated' | 'deleted' | 'unchanged' | 'skipped';

export interface DreamCollectorRun {
  runId: string;
  scope: DreamScope;
  sourceSession?: DreamSourceSessionRef;
  status: DreamCollectorRunStatus;
  decision?: DreamCollectorDecision;
  dreamIds?: string[];
  startedAtUnixMs: number;
  completedAtUnixMs?: number;
  backoffUntilUnixMs?: number;
  error?: string;
  logPath?: string;
}

export type DreamLogLevel = 'debug' | 'info' | 'warn' | 'error';

export interface DreamLogEntry {
  timeUnixMs: number;
  level: DreamLogLevel;
  event: string;
  runId?: string;
  scope?: DreamScope;
  sourceSessionPath?: string;
  message?: string;
  data?: JsonObject;
}

export interface DreamSettings {
  enabled: boolean;
  collectorModel?: string;
  debugMode: boolean;
}

export interface ChatArchive {
  messages: Array<{ role: 'user' | 'assistant'; content: string }>;
  assistantAux: AssistantAuxArchiveEntry[];
  llmHistory: Array<StoredLlmMessageArchiveEntry | LegacyLlmMessageArchiveEntry>;
  subagentSessions?: SubagentSessionArchiveEntry[];
  loopEnabled?: boolean;
  approvalLevel?: 'default' | 'full-approval';
}

export type McpStatusState = 'idle' | 'loading' | 'ready' | 'error';

export interface McpStatusSnapshot {
  revision: number;
  state: McpStatusState;
  configuredServers: number;
  loadedServers: number;
  cachedTools: number;
  lastError?: string;
}

export interface AppPaths {
  workspaceRoot(): string;
  configFile(): string;
  chatsDir(): string;
  permissionsFile(): string;
  logFile(): string;
}

export interface Telemetry {
  logEvent(message: string): void;
  logJsonHttpBody(label: string, payload: JsonValue): void;
}

export interface ConfigStore<Config> {
  load(): Promise<Config>;
  save(config: Config): Promise<void>;
}

export interface SecretStore {
  loadGlobalApiKey(): Promise<string | undefined>;
  saveGlobalApiKey(apiKey: string): Promise<void>;
  removeGlobalApiKey(): Promise<void>;
  loadModelApiKey(modelName: string): Promise<string | undefined>;
  saveModelApiKey(modelName: string, apiKey: string): Promise<void>;
  removeModelApiKey(modelName: string): Promise<void>;
  hasModelApiKey(modelName: string): Promise<boolean>;
}

export interface ChatRepository {
  list(): Promise<string[]>;
  save(path: string | undefined, archive: ChatArchive): Promise<string>;
  load(path: string): Promise<ChatArchive>;
}

export interface ToolCallRequest {
  id: string;
  name: string;
  argumentsJson: string;
}

export type ToolAgentStep =
  | { kind: 'tool-calls'; calls: ToolCallRequest[] }
  | { kind: 'final-response-ready' };

export interface LlmTokenUsage {
  inputTokens: number;
  outputTokens: number;
  totalTokens: number;
  reasoningTokens?: number;
  cachedInputTokens?: number;
}

export interface ToolAgentRoundResult<State = JsonValue> {
  state: State;
  step: ToolAgentStep;
  requestTrace: JsonValue[];
  usage?: LlmTokenUsage;
  /** Provider SDK executed built-in search in-stream but the answer step is not ready yet. */
  resumeStreamingAfterProviderSearch?: boolean;
}

export type ToolAgentRoundCompletion<State = JsonValue> =
  | { kind: 'success'; result: ToolAgentRoundResult<State> }
  | { kind: 'failure'; error: string; requestTrace: JsonValue[] };

export type LlmStreamEvent =
  | { kind: 'thinking-chunk'; text: string }
  | {
      kind: 'streaming-tool-preview';
      toolCallId: string;
      toolName: string;
      argumentsJson: string;
    }
  | { kind: 'assistant-chunk'; text: string }
  | {
      kind: 'history-compacted';
      newHistory: LlmMessage[];
      droppedMessages: number;
    }
  | { kind: 'done' }
  | { kind: 'error'; error: string };

export interface StartedToolAgentRound<State = JsonValue> {
  eventStream: AsyncIterable<LlmStreamEvent>;
  completion: Promise<ToolAgentRoundCompletion<State>>;
  cancel?: () => void;
}

export type AskQuestionsQuestionKind = 'single_select' | 'multi_select' | 'text';

export interface AskQuestionsOption {
  label: string;
  summary?: string;
}

export interface AskQuestionsQuestion {
  id: string;
  title: string;
  kind: AskQuestionsQuestionKind;
  required?: boolean;
  options?: AskQuestionsOption[];
  allowCustomInput?: boolean;
  customInputPlaceholder?: string;
  customInputLabel?: string;
}

export interface AskQuestionsRequest {
  title?: string;
  questions: AskQuestionsQuestion[];
}

export interface AskQuestionsAnswer {
  questionId: string;
  title: string;
  kind: AskQuestionsQuestionKind;
  answered: boolean;
  selectedOptionIndexes?: number[];
  selectedOptionLabels?: string[];
  customInput?: string;
  text?: string;
}

export type AskQuestionsResult =
  | { status: 'skipped' }
  | { status: 'answered'; answers: AskQuestionsAnswer[] };

export interface RunSubagentRequest {
  task: string;
  successCriteria?: string;
  contextSummary?: string;
  filesToInspect?: string[];
  expectedOutput?: string;
}

export type AuthorizationDecision<TrustTarget = string> =
  | { kind: 'allowed' }
  | { kind: 'need-approval'; prompt: string; trustTarget?: TrustTarget }
  | { kind: 'need-questions'; questions: AskQuestionsRequest };

export interface ToolRequestExecutionMetadata {
  toolCallId?: string;
  toolName?: string;
  subagentSessionId?: string;
  subagentTitle?: string;
  userInitiated?: boolean;
}

export interface ToolExecutor<
  ToolRequest = unknown,
  TrustTarget = string,
  McpServer = unknown,
  McpTool = unknown,
  McpResource = unknown,
  McpPrompt = unknown,
  McpInspection = unknown,
> {
  toolDefinitionsJson(): JsonValue;
  /** When Loop is off, omit finish_task from toolDefinitionsJson(). */
  setLoopToolExposure?(loopEnabled: boolean): void;
  setAgentModeToolExposure?(agentMode: SpiritAgentMode): void;
  /** @deprecated Use {@link setAgentModeToolExposure}. */
  setPlanModeToolExposure?(planMode: boolean): void;
  parseCommand(message: string): Promise<ToolRequest>;
  requestFromFunctionCall(name: string, argumentsJson: string): Promise<ToolRequest>;
  authorize(request: ToolRequest): Promise<AuthorizationDecision<TrustTarget>>;
  trust(target: TrustTarget): Promise<void>;
  execute(request: ToolRequest): Promise<ToolExecutionOutput>;
  attachRequestMetadata?(request: ToolRequest, metadata: ToolRequestExecutionMetadata): ToolRequest;
  continueAfterQuestions?(request: ToolRequest, result: AskQuestionsResult): Promise<ToolRequest | undefined>;
  shouldExecuteInBackground?(request: ToolRequest): boolean;
  backgroundStatusText?(request: ToolRequest): string | undefined;
  startMcpBackgroundRefresh(): void;
  mcpStatusSnapshot(): McpStatusSnapshot;
  addMcpServer(name: string, config: JsonValue): Promise<string>;
  listMcpServers(): Promise<McpServer[]>;
  inspectMcpServer(name: string): Promise<McpInspection>;
  listMcpTools(name: string): Promise<McpTool[]>;
  listMcpResources(name: string): Promise<McpResource[]>;
  readMcpResource(name: string, uri: string): Promise<JsonValue>;
  listCachedMcpPrompts(name: string): Promise<McpPrompt[]>;
  listMcpPrompts(name: string): Promise<McpPrompt[]>;
  getMcpPrompt(name: string, prompt: string, argsJson?: string): Promise<JsonValue>;
}

export interface LlmTransport<Config, State = JsonValue> {
  startToolAgentRound(
    config: Config,
    state: State,
    tools: JsonValue,
  ): Promise<ToolAgentRoundCompletion<State>>;
  startToolAgentRoundStreaming?(
    config: Config,
    state: State,
    tools: JsonValue,
  ): Promise<StartedToolAgentRound<State>>;
  compactHistoryManual(
    config: Config,
    history: LlmMessage[],
    onProgress?: (message: string) => void,
  ): Promise<{
    droppedMessages: number;
    beforeLength: number;
    afterLength: number;
  }>;
  compactSummaryText(history: LlmMessage[]): string | undefined;
  isContextOverflowError(error: string): boolean;
  llmHistoryAsApiMessages(history: LlmMessage[]): JsonValue[];
  llmSystemPromptsForExport(): JsonValue;
}

File: apps/cli/src/ports.rs
use anyhow::Result;
use serde::{Deserialize, Deserializer, Serialize};
use serde_json::Value;
use std::path::PathBuf;

use crate::model_registry::AppConfig;

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct AssistantAuxArchiveEntry {
    pub message_index: usize,
    pub thinking: Option<String>,
    pub compaction: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub finish_task_notice: Option<String>,
}

#[derive(Clone, Copy, Debug, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum SubagentSessionStatus {
    Running,
    Completed,
    Failed,
    Blocked,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentSessionSummary {
    pub session_id: String,
    pub parent_tool_call_id: String,
    pub title: String,
    pub status: SubagentSessionStatus,
    pub started_at_unix_ms: u64,
    pub updated_at_unix_ms: u64,
    pub completed_at_unix_ms: Option<u64>,
    pub latest_message: Option<String>,
    pub final_output: Option<String>,
    pub error: Option<String>,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(tag = "type", rename_all = "camelCase")]
pub enum LlmContentPart {
    Text { text: String },
    Image { path: String },
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ArchivedLlmToolCall {
    pub id: String,
    pub name: String,
    pub arguments_json: String,
}

#[derive(Clone, Debug, Serialize)]
#[serde(rename_all = "camelCase")]
pub struct ArchivedLlmMessage {
    pub role: String,
    pub content: Vec<LlmContentPart>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub tool_call_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub tool_calls: Option<Vec<ArchivedLlmToolCall>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub provider_state: Option<Value>,
}

impl ArchivedLlmMessage {
    pub fn from_text_and_images(role: String, content: String, image_paths: Vec<String>) -> Self {
        let mut parts = Vec::new();
        if !content.is_empty() {
            parts.push(LlmContentPart::Text { text: content });
        }
        for path in image_paths {
            parts.push(LlmContentPart::Image { path });
        }
        Self {
            role,
            content: parts,
            tool_call_id: None,
            tool_calls: None,
            provider_state: None,
        }
    }

    pub fn with_tool_call_id(mut self, tool_call_id: Option<String>) -> Self {
        self.tool_call_id = tool_call_id;
        self
    }

    pub fn with_tool_calls(mut self, tool_calls: Option<Vec<ArchivedLlmToolCall>>) -> Self {
        self.tool_calls = tool_calls;
        self
    }

    pub fn with_provider_state(mut self, provider_state: Option<Value>) -> Self {
        self.provider_state = provider_state;
        self
    }

    pub fn text_content(&self) -> String {
        self.content
            .iter()
            .filter_map(|part| match part {
                LlmContentPart::Text { text } => Some(text.as_str()),
                LlmContentPart::Image { .. } => None,
            })
            .collect::<String>()
    }

    pub fn image_paths(&self) -> Vec<String> {
        self.content
            .iter()
            .filter_map(|part| match part {
                LlmContentPart::Image { path } => Some(path.clone()),
                LlmContentPart::Text { .. } => None,
            })
            .collect()
    }
}

impl<'de> Deserialize<'de> for ArchivedLlmMessage {
    fn deserialize<D>(deserializer: D) -> std::result::Result<Self, D::Error>
    where
        D: Deserializer<'de>,
    {
        #[derive(Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct CurrentArchivedLlmMessage {
            role: String,
            content: Vec<LlmContentPart>,
            #[serde(default, alias = "tool_call_id")]
            tool_call_id: Option<String>,
            #[serde(default, alias = "toolCalls")]
            tool_calls: Option<Vec<ArchivedLlmToolCall>>,
            #[serde(default, alias = "providerState")]
            provider_state: Option<Value>,
        }

        #[derive(Deserialize)]
        #[serde(rename_all = "camelCase")]
        struct LegacyArchivedLlmMessage {
            role: String,
            content: String,
            #[serde(default)]
            image_paths: Vec<String>,
            #[serde(default, alias = "tool_call_id")]
            tool_call_id: Option<String>,
            #[serde(default, alias = "toolCalls")]
            tool_calls: Option<Vec<ArchivedLlmToolCall>>,
            #[serde(default, alias = "providerState")]
            provider_state: Option<Value>,
        }

        #[derive(Deserialize)]
        #[serde(untagged)]
        enum ArchivedLlmMessageRepr {
            Current(CurrentArchivedLlmMessage),
            Legacy(LegacyArchivedLlmMessage),
        }

        match ArchivedLlmMessageRepr::deserialize(deserializer)? {
            ArchivedLlmMessageRepr::Current(message) => Ok(Self {
                role: message.role,
                content: message.content,
                tool_call_id: message.tool_call_id,
                tool_calls: message.tool_calls,
                provider_state: message.provider_state,
            }),
            ArchivedLlmMessageRepr::Legacy(message) => {
                Ok(
                    Self::from_text_and_images(message.role, message.content, message.image_paths)
                        .with_tool_call_id(message.tool_call_id)
                        .with_tool_calls(message.tool_calls)
                        .with_provider_state(message.provider_state),
                )
            }
        }
    }
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct SubagentSessionArchiveEntry {
    pub summary: SubagentSessionSummary,
    #[serde(default)]
    pub llm_history: Vec<ArchivedLlmMessage>,
}

fn default_approval_level() -> String {
    "default".to_string()
}

/// Canonical approval levels shown in CLI help / unknown-value errors.
pub const APPROVAL_LEVELS: [&str; 3] = ["default", "auto-approval", "bypass-approval"];

pub fn available_approval_levels_csv() -> String {
    APPROVAL_LEVELS.join(", ")
}

/// Soft normalize used by archives / internal paths: unknown values become `default`.
pub fn normalize_approval_level(value: &str) -> String {
    parse_approval_level_strict(value).unwrap_or_else(|| "default".to_string())
}

/// Strict parse for CLI flags: unknown values return `None` (do not fall back to default).
pub fn parse_approval_level_strict(value: &str) -> Option<String> {
    match value.trim() {
        "default" => Some("default".to_string()),
        "auto-approval" => Some("auto-approval".to_string()),
        "bypass-approval" => Some("bypass-approval".to_string()),
        _ => None,
    }
}

pub fn normalize_llm_http_version(value: &str) -> String {
    let normalized = value.trim().to_ascii_lowercase();
    if normalized == "http1.1" || normalized == "http/1.1" || normalized == "http1" {
        "http1.1".to_string()
    } else {
        "http2".to_string()
    }
}

pub const TUI_MODE_INLINE: &str = "inline";
pub const TUI_MODE_FULLSCREEN: &str = "fullscreen";
pub const TUI_MODES: [&str; 2] = [TUI_MODE_INLINE, TUI_MODE_FULLSCREEN];

pub fn available_tui_modes_csv() -> String {
    TUI_MODES.join(", ")
}

/// Strict parse for CLI flags and `/tui`: unknown values return `None`.
pub fn parse_tui_mode_strict(value: &str) -> Option<String> {
    match value.trim().to_ascii_lowercase().as_str() {
        TUI_MODE_INLINE => Some(TUI_MODE_INLINE.to_string()),
        TUI_MODE_FULLSCREEN => Some(TUI_MODE_FULLSCREEN.to_string()),
        _ => None,
    }
}

/// Soft normalize used by config: unknown values become `inline`.
pub fn normalize_tui_mode(value: &str) -> String {
    parse_tui_mode_strict(value).unwrap_or_else(|| TUI_MODE_INLINE.to_string())
}

#[derive(Clone, Copy, Debug, PartialEq, Eq)]
pub enum AttachChatSessionOutcome {
    /// `session.attach` missed; created a session and hydrated from disk archive.
    Created,
    /// Joined an existing live daemon session for this conversation key.
    AttachedLive,
}

#[derive(Clone, Debug, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct ChatArchive {
    pub messages: Vec<(String, String)>,
    pub assistant_aux: Vec<AssistantAuxArchiveEntry>,
    pub llm_history: Vec<ArchivedLlmMessage>,
    #[serde(default)]
    pub loop_enabled: bool,
    #[serde(default = "default_approval_level")]
    pub approval_level: String,
    #[serde(default)]
    pub subagent_sessions: Vec<SubagentSessionArchiveEntry>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub desktop_messages: Option<Vec<crate::rewind::ConversationMessageSnapshot>>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub rewind: Option<serde_json::Value>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub session_display_name: Option<String>,
}

#[derive(Clone, Copy, Debug, Default, PartialEq, Eq, Serialize, Deserialize)]
#[serde(rename_all = "lowercase")]
pub enum McpStatusState {
    #[default]
    Idle,
    Loading,
    Ready,
    Error,
}

#[derive(Clone, Debug, Default, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct McpStatusSnapshot {
    pub revision: u64,
    pub state: McpStatusState,
    pub configured_servers: usize,
    pub loaded_servers: usize,
    pub cached_tools: usize,
    pub last_error: Option<String>,
}

pub trait AppPaths: Send + Sync {
    fn workspace_root(&self) -> PathBuf;
    fn config_file(&self) -> PathBuf;
    fn chats_dir(&self) -> PathBuf;
    fn log_file(&self) -> PathBuf;
}

pub trait ConfigStore: Send + Sync {
    fn load(&self) -> Result<AppConfig>;
    fn save(&self, config: &AppConfig) -> Result<()>;
}

pub trait SecretStore: Send + Sync {
    fn load_global_api_key(&self) -> Result<Option<String>>;
    fn save_global_api_key(&self, api_key: &str) -> Result<()>;
    fn remove_global_api_key(&self) -> Result<()>;
    fn load_model_api_key(&self, model_name: &str) -> Result<Option<String>>;
    fn save_model_api_key(&self, model_name: &str, api_key: &str) -> Result<()>;
    fn remove_model_api_key(&self, model_name: &str) -> Result<()>;
    fn has_model_api_key(&self, model_name: &str) -> Result<bool>;
}

#[derive(Clone, Debug)]
pub struct ChatSessionListItem {
    pub path: String,
    pub display_name: String,
    pub modified_at_unix_ms: u128,
}

pub trait ChatRepository: Send + Sync {
    fn list(&self) -> Result<Vec<ChatSessionListItem>>;
    fn save(&self, path: Option<&str>, archive: &ChatArchive) -> Result<PathBuf>;
    fn load(&self, path: &str) -> Result<ChatArchive>;
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn parse_approval_level_strict_accepts_canonical_values() {
        assert_eq!(
            parse_approval_level_strict("default").as_deref(),
            Some("default")
        );
        assert_eq!(
            parse_approval_level_strict("auto-approval").as_deref(),
            Some("auto-approval")
        );
        assert_eq!(
            parse_approval_level_strict("bypass-approval").as_deref(),
            Some("bypass-approval")
        );
    }

    #[test]
    fn parse_approval_level_strict_rejects_unknown() {
        assert_eq!(parse_approval_level_strict("bogus"), None);
        assert_eq!(parse_approval_level_strict(""), None);
        assert_eq!(parse_approval_level_strict("full-approval"), None);
        assert_eq!(parse_approval_level_strict("auto"), None);
    }

    #[test]
    fn available_approval_levels_csv_uses_comma_space() {
        assert_eq!(
            available_approval_levels_csv(),
            "default, auto-approval, bypass-approval"
        );
    }

    #[test]
    fn normalize_approval_level_falls_back_unknown_to_default() {
        assert_eq!(normalize_approval_level("nope"), "default");
    }

    #[test]
    fn parse_tui_mode_strict_accepts_canonical() {
        assert_eq!(parse_tui_mode_strict("inline").as_deref(), Some("inline"));
        assert_eq!(
            parse_tui_mode_strict("FULLSCREEN").as_deref(),
            Some("fullscreen")
        );
    }

    #[test]
    fn parse_tui_mode_strict_rejects_unknown() {
        assert_eq!(parse_tui_mode_strict("mini"), None);
        assert_eq!(parse_tui_mode_strict(""), None);
        assert_eq!(parse_tui_mode_strict("window"), None);
    }

    #[test]
    fn available_tui_modes_csv_uses_comma_space() {
        assert_eq!(available_tui_modes_csv(), "inline, fullscreen");
    }

    #[test]
    fn normalize_tui_mode_falls_back_unknown_to_inline() {
        assert_eq!(normalize_tui_mode("nope"), "inline");
        assert_eq!(normalize_tui_mode("fullscreen"), "fullscreen");
    }
}
